text/3830-dedented-string-literals.md

```diff
@@ -0,0 +1,1296 @@
+
+```rs
+fn main() {
+    println!(d"
+        create table student(
+            id int primary key,
+            name text
+        )"); // ERROR
+}
+```
+
+- Having the closing quote **always** be on its own line makes it more obvious to the reader from which column onwards leading indentation will be removed.
+- In the example above, it is not immediately clear where that would be from.
+- It easy to modify the common indentation level of the string in the future, as you do not have to create a new line.
+
+## Differences from RFC 3450
+
+The [RFC #3450: Propose code string literals](https://github.com/rust-lang/rfcs/pull/3450) is similar to this one, however this RFC is different and this section explains why.
+
+Differences:
+
+- #3450 uses `h` as the modifier instead of `d`.
+
+    proposes using `h` as acronym for [Here document](https://en.wikipedia.org/wiki/Here_document).
+
+    The term is likely to be less known, and may raise confusion.
+
+    Additionally, here documents are more associated with "code blocks". While this feature is useful for code blocks, it is not just for them.
+
+    While the `d` mnemonic for **dedent** clearly describes what actually happens to the strings.
+
+- #3450 allows to write an *info string*, like in markdown.
+
+    It proposes the ability to write:
+
+    ```rs
+    let sql = d"sql
+        SELECT * FROM table;
+        ";
+    ```
+
+    With the `sql` not affecting the output, but can aid in syntax highlighting and such.
+
+   1. This is not necessary, as at the moment you can add a block comment next to the string, which syntax highlighters can use *today* to inject whatever language is specified.
+ 
+         ```rs
+         let sql = /* sql */ "SELECT * FROM table;";
+         ```
+ 
+   2. Is considered out of scope for this RFC to consider.
+ 
+         It would be a backward-compatible change to make for a future RFC, if it's desired.
+
+   3. [Expression attributes](https://github.com/rust-lang/rust/issues/15701) are likely to be more suitable for this purpose. (not part of this RFC)
+
+         ```rs
+         let sql = #[editor::language("sql")] "SELECT * FROM table;";
+         ```
+
+- RFC #3450 makes the "code strings" always end with a newline, with the ability to prepend a minus before the closing quote in order to remove the final newline.
+
+    However, in this RFC the following:
+
+    ```rs
+    print!(d"
+        a
+        ");
+    ^^^^ // common leading whitespace (will be removed)
+    ```
+
+    Prints: `a`
+
+    **Without** an ending newline.
+
+    In order to add a newline at the end, you have to add a newline in the source code:
+
+    ```rs
+    print!(d"
+        a
+
+        ");
+    ^^^^ // common leading whitespace (will be removed)
+    ```
+
+    The above prints:
+
+    ```
+    a  
+    ```
+
+    **With** a newline.
+
+    Additionally, finishing with `-"` instead of `"` is not seen anywhere in the language, and would not fit in.
+
+## Use a macro instead
```

Comment:
This section seems to omit the obvious option: what about a built-in macro? Some of the potential downsides like having to wrap the string in a macro invocation could be mitigated by [postfix macros](https://github.com/rust-lang/rfcs/pull/2442), which also have an open RFC.

Plus, as I also comment on later down, a built-in macro mitigates most of the performance issues of a separate proc macro.